Zend/zend_observer.c

```diff
@@ -0,0 +1,161 @@
+	zval *return_value);
+
+// Call during minit/startup ONLY
+ZEND_API void zend_observer_fcall_register(zend_observer_fcall_init init) {
+	/* We don't want to get an extension handle unless an ext installs an observer */
+	if (!ZEND_OBSERVER_ENABLED) {
+		zend_observer_fcall_op_array_extension =
+			zend_get_op_array_extension_handle();
+
+		/* ZEND_CALL_TRAMPOLINE has SPEC(OBSERVER) but zend_init_call_trampoline_op()
+		 * is called before any extensions have registered as an observer. So we
+		 * adjust the offset to the observed handler when we know we need to observe. */
+		ZEND_VM_SET_OPCODE_HANDLER(&EG(call_trampoline_op));
+
+		/* ZEND_HANDLE_EXCEPTION also has SPEC(OBSERVER) and no observer extensions
+		 * exist when zend_init_exception_op() is called. */
+		ZEND_VM_SET_OPCODE_HANDLER(EG(exception_op));
+		ZEND_VM_SET_OPCODE_HANDLER(EG(exception_op)+1);
+		ZEND_VM_SET_OPCODE_HANDLER(EG(exception_op)+2);
+	}
+	zend_llist_add_element(&zend_observers_fcall_list, &init);
+}
+
+// Called by engine before MINITs
+ZEND_API void zend_observer_startup(void) {
+	zend_llist_init(&zend_observers_fcall_list, sizeof(zend_observer_fcall_init), NULL, 1);
+}
+
+ZEND_API void zend_observer_activate(void) {
+	/* todo: how to size the arena? */
+	if (ZEND_OBSERVER_ENABLED) {
+		fcall_handlers_arena = zend_arena_create(1024);
+	}
+}
+
+ZEND_API void zend_observer_deactivate(void) {
+	if (fcall_handlers_arena) {
+		zend_arena_destroy(fcall_handlers_arena);
+	}
+}
+
+ZEND_API void zend_observer_shutdown(void) {
+	zend_llist_destroy(&zend_observers_fcall_list);
+}
+
+ZEND_API void zend_observer_fcall_install(zend_function *function) {
+	zend_llist_element *element;
+	zend_llist *list = &zend_observers_fcall_list;
+	zend_op_array *op_array = &function->op_array;
+
+	if (fcall_handlers_arena == NULL) {
+		return;
+	}
+
+	ZEND_ASSERT(function->type != ZEND_INTERNAL_FUNCTION);
+
+	zend_llist handlers_list;
+	zend_llist_init(&handlers_list, sizeof(zend_observer_fcall), NULL, 0);
+	for (element = list->head; element; element = element->next) {
+		zend_observer_fcall_init init;
+		memcpy(&init, element->data, sizeof init);
+		zend_observer_fcall handlers = init(function);
+		if (handlers.begin || handlers.end) {
+			zend_llist_add_element(&handlers_list, &handlers);
+		}
+	}
+
+	ZEND_ASSERT(RUN_TIME_CACHE(op_array));
+	void *ext;
+	if (handlers_list.count) {
+		size_t size = sizeof(zend_observer_fcall_cache) + (handlers_list.count - 1) * sizeof(zend_observer_fcall);
+		zend_observer_fcall_cache *cache = zend_arena_alloc(&fcall_handlers_arena, size);
+		zend_observer_fcall *handler = cache->handlers;
+		for (element = handlers_list.head; element; element = element->next) {
+			memcpy(handler++, element->data, sizeof *handler);
+		}
+		cache->end = handler;
+		ext = cache;
+	} else {
+		ext = ZEND_OBSERVER_NOT_OBSERVED;
+	}
+
+	ZEND_OBSERVER_HANDLERS(op_array) = ext;
+	zend_llist_destroy(&handlers_list);
+}
+
+void zend_observe_fcall_begin(
```

Comment:
Do the definitions of `zend_observe_fcall_begin`, `zend_observer_fcall_call_end_helper`, and `zend_observe_fcall_end` need `ZEND_API` too? Or is it sufficient for the declaration only? I don't know Windows, which mostly seems to be the only platform this affects.

Reply:
I don't see a reason to export them at the moment. We can always export them later if there is a need to. That being said, we could probably remove `ZEND_API` from the startup/activate and shutdown/deactivate functions. Not sure those should ever be called from a linked lib.